### PR TITLE
HDkeyring

### PR DESCRIPTION
**Pull request by darkdh**
Resolves https://github.com/brave/brave-browser/issues/15215

`EthAddress`, `EthTransaction` and recovery id of `HDKey::Sign` are prerequisites of `HDKeyring` implementation

EthAddress support checksum address (EIP55 + EIP1191)
EthTransaction support EIP155, adding chain_id to `GetMessageToSign` and `ProcessSignature`

`HDKey::GetUncompressedPublicKey` is needed to generate the eth address
Public key recovery is for generate v, r, s for transaction signature

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by darkdh** [comment]
@orspetol 
This is the commit that added `secp256k1_recovery` 
https://github.com/brave/brave-core/pull/8475/commits/17ec77f42be505817792cad187a0c1ef01a9ad4e

File: components/brave_wallet/browser/hd_key.cc [hunk 1]
```diff
@@ -166,6 +167,25 @@ std::string HDKey::GetPublicExtendedKey() const {
   return Serialize(MAINNET_PUBLIC, public_key_);
 }
 
+std::vector<uint8_t> HDKey::GetUncompressedPublicKey() const {
+  // uncompressed
```
Comment (orspetol) [inline components/brave_wallet/browser/hd_key.cc:171]: Will we support compressed? Just curious why one and not both?

Comment (darkdh) [reply components/brave_wallet/browser/hd_key.cc:171]: HDKey stores compressed public key and we don't need to access that directly, we provide `GetPublicExtendedKey` instead which is base58(hash160(public_key) 
Uncompressed public key is used only for eth address calculation. We want to restrict the access to `secp256k1_context` in the scope of HDKey so we don't want callers to get a compressed public key and covert it on their own.

**Review by orspetol — APPROVED** [review]

**Review by bbondy — APPROVED** [review]
Great work!

File: components/brave_wallet/browser/eth_address.cc [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+  const std::string hash_str(KeccakHash(input).data() + 2);
+  const std::string address_str =
+      base::ToLowerASCII(base::HexEncode(bytes_.data(), bytes_.size()));
+
+  for (size_t i = 0; i < address_str.length(); ++i) {
+    if (isdigit(address_str[i])) {
+      result.push_back(address_str[i]);
+    } else {  // address has already be validated
+      int nibble = std::stoi(std::string(1, hash_str[i]), nullptr, 16);
```
Comment (bridiver) [inline components/brave_wallet/browser/eth_address.cc:81]: https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++-features.md#string_number-conversion-functions-banned
I know there are some usages in upstream, but I assume those are mistakes

Comment (darkdh) [reply components/brave_wallet/browser/eth_address.cc:81]: yep, should have used `base/strings/string_number_conversions.h `

**Pull request merged**